pkg/hub/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,230 @@
+		return m.manifestWorkLister.ManifestWorks(ns.Name).List(labels.Everything())
+	})
+
+	if err != nil {
+		klog.Errorf("Reconcile role/rolebinding %s fails with err: %v", key, err)
+	}
+	return err
+}
+
+func (m *FinalizeController) syncRoleAndRoleBinding(ctx context.Context, controllerContext factory.SyncContext,
+	role *rbacv1.Role, rolebinding *rbacv1.RoleBinding, ns *corev1.Namespace, listWorksFunc listManifestWorksFunc) error {
+	klog.V(4).Infof("Sync role/rolebinding %s/%s%s", ns.Name, ns.Name, keySuffix)
+	// no work if neither role nor rolebinding exists
+	if role == nil && rolebinding == nil {
+		return nil
+	}
+
+	// no work if neither role nor rolebinding has the finalizer
+	if !hasFinalizer(role, rolebinding) {
+		return nil
+	}
+
+	if ns.DeletionTimestamp == nil || ns.DeletionTimestamp.IsZero() {
+		// skip the role if it is in a non-terminating namespace and not being deleted
+		if role != nil && (role.DeletionTimestamp == nil || role.DeletionTimestamp.IsZero()) {
+			role = nil
+		}
+
+		// skip the rolebinding if it is in a non-terminating namespace and not being deleted
+		if rolebinding != nil && (rolebinding.DeletionTimestamp == nil || rolebinding.DeletionTimestamp.IsZero()) {
+			rolebinding = nil
+		}
+	} else {
+		works, err := listWorksFunc()
+		if err != nil {
+			return err
+		}
+
+		// requeue the key if there exists any manifest work in cluster namespace
+		if len(works) != 0 {
+			key := controllerContext.QueueKey()
+			controllerContext.Queue().AddAfter(key, RequeueDelay)
+			klog.V(4).Infof("Requeue role/rolebinding %q after %v", key, RequeueDelay)
+			return nil
+		}
+	}
+
+	// remove finalizer from role/rolebinding
+	errs := []error{}
+	if role != nil {
+		role = role.DeepCopy()
+		if changed := removeFinalizer(role, manifestWorkFinalizer); changed {
+			_, err := m.rbacClient.Roles(role.Namespace).Update(ctx, role, metav1.UpdateOptions{})
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if rolebinding != nil {
+		rolebinding = rolebinding.DeepCopy()
+		if changed := removeFinalizer(rolebinding, manifestWorkFinalizer); changed {
+			_, err := m.rbacClient.RoleBindings(rolebinding.Namespace).Update(ctx, rolebinding, metav1.UpdateOptions{})
+			if err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+
+	if len(errs) > 0 {
+		return utilerrors.NewAggregate(errs)
+	}
+	return nil
+}
+
+func (m *FinalizeController) getRoleAndRoleBinding(namespace, name string) (*rbacv1.Role, *rbacv1.RoleBinding, error) {
+	role, err := m.roleLister.Roles(namespace).Get(name)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, nil, err
+	}
+
+	rolebinding, err := m.roleBindingLister.RoleBindings(namespace).Get(name)
+	if err != nil && !errors.IsNotFound(err) {
+		return nil, nil, err
+	}
+
+	return role, rolebinding, nil
+}
+
+func hasFinalizer(role *rbacv1.Role, rolebinding *rbacv1.RoleBinding) bool {
```

Comment:
update this to be a function on ObjectMeta and then call it twice please.  I have trouble reading the code.